solutions/usaco-863.mdx

```diff
@@ -2,17 +2,60 @@
 id: usaco-863
 source: USACO Gold 2018 December
 title: Teamwork
-author: Sofia Yang
+author: Sofia Yang, Ryan Chou
 ---
 
 [Official Analysis (C++)](http://www.usaco.org/current/data/sol_teamwork_gold_dec18.html)
 
+## Explanation
+
+The important observation to make is that we don't need to store the groupings of cows, just the best possible sum so far. This hints at a [DP](/gold/intro-dp). Let $\texttt{dp[i]} =$ the maximum skill level for cows up to $i$, then we'll update $\texttt{dp[i]}$ for every previous cow down to $i - k + 1$, where $k$ represents the team size.
 ## Implementation
 
 **Time Complexity:** $\mathcal{O}(N \cdot K)$
 
 <LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+int main() {
+	freopen("teamwork.in", "r", stdin);
+	freopen("teamwork.out", "w", stdout);
+
+	int n, k;
+	cin >> n >> k;
+
+	vector<int> skill(n);
+	for (int i = 0; i < n; i++) {
+		cin >> skill[i];
+	}
+
+	vector<int> dp(n, -1);
+
+	for (int i = 0; i < n; i++) {
+		// by not joining it to a team, the default value is just skill[i]
+		int cur = skill[i];
 
+		for (int j = i; j >= (i - k + 1) && ~j; j--) {
+			cur = max(cur, skill[j]);
+			/*
+			 * update answer: dp[i] = max(dp[i], 
+			 * the skill of joining this cow to a team + all of the previous sums)
+			 */
+			if (j > 0) {
```

Comment:
if this is supposed to check for just -1, be more explicit
```suggestion
			if (j > -1) {
```

Reply:
it's checking whether it's greater than 0...

if it's equal to zero, then we don't subtract one, or else it'll go out of bounds.

Reply:
ok cool